### PR TITLE
DATAUP-487 Staging Area Viewer reworking to allow uploading of all checked files

### PR DESCRIPTION
**Pull request by briehl**
# Description of PR purpose/changes

This has a number of changes to the staging area viewer code to support some bug fixes. There's a few other changes that this included.

Sorry for the size, there was really no way to do this piecemeal, as it needed an overhaul of how the datatable gets created and maintained. I couldn't really see a way to do that in parts without introducing strange and unusual behavior.

Bugs addressed:
* Navigating back from a two-deep-nested directory used to go from, for example `user/dir1/dir2` to `user/user/dir1`. Now it goes to `user/dir1` as expected.
* Selected files from multiple pages now all get added to the import cell.
* Select all / select none button works again.
* Selected files from nested directories all get added to the import cell.
  * For example, selecting a file from root, a file from one subdirectory, and files from a deeper directory, will all get added to the bulk import cell
* File checked state persists through refreshes.

Structural / architecture changes:
* The `render` function should now only get run once. This builds the DataTable, which should also only be done once.
* Previously, we used data built into the DOM with a template file and applied the DataTable to that. Now, we just feed the data rows into it.
* The DataTable now knows how to render the data from each file object. On refresh, this fetches the data again from the server, and just feeds it into the DataTable.
* This now fetches all the files from the Staging Area and just feeds them to the table, which filters based on what's the current directory. This really isn't that much of a change, as the `StagingArea.list()` function worked recursively. So running `list()` on the root directory returned all files anyway. We now just take advantage of that instead of throwing things away.
* Each file is now treated as unique based on its path, not its file name. If there are files with the same name in multiple directories, this should now be ok - before it led to Strange Behavior.

# Jira Ticket / Issue #
e.g. https://kbase-jira.atlassian.net/browse/DATAUP-487
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
* Throw some files in your staging area, enough to show multiple pages / directories
* Toggle through pages and select files
* Do the import step, and watch them all appear!
* Also, try a refresh, deletion, download, or other function to make sure nothing was lost
- [x] Tests pass locally and in GitHub Actions
- [x] Changes available by spinning up a local narrative and navigating to _X_ to see _Y_

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook

### CODE REVIEW

File: kbase-extension/static/kbase/templates/data_staging/ftp_file_table.html [hunk 1]
```diff
@@ -13,54 +13,11 @@
         </tr>
     </thead>
     <tbody class="kb-staging-table-body">
-    {{#each files}}
```
Comment (briehl) [inline kbase-extension/static/kbase/templates/data_staging/ftp_file_table.html:16]: This all gets done directly by DataTables now. Might consider doing away with the whole template in the future.

File: test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js [hunk 2]
```diff
@@ -3,18 +3,20 @@ define([
...
         const startingPath = '/',
-            updatePathFn = null, //() => {},
-            fakeUser = 'notAUser';
+            updatePathFn = null,
+            fakeUser = 'notAUser',
+            stagingServiceUrl = Config.url('staging_api_url');
 
         beforeEach(() => {
             jasmine.Ajax.install();
-            jasmine.Ajax.stubRequest(/.*\/staging_service\/list\/?/).andReturn({
```
Comment (briehl) [inline test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js:17]: Set this to be configured, as I was running a Staging Service locally and this was trying to pass calls through!

File: test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js [hunk 3]
```diff
@@ -26,29 +28,45 @@ define([
                         mtime: 1532738637499,
                         size: 34,
                         isFolder: true,
-                        mappings: null,
```
Comment (briehl) [inline test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js:29]: `mappings` isn't in the file list results

File: test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js [hunk 4]
```diff
@@ -26,29 +28,45 @@ define([
...
                 ]),
             });
 
-            const mappings = { mappings: [null, null, null] };
-            jasmine.Ajax.stubRequest(/.*\/staging_service\/importer_mappings\/?/).andReturn({
+            const mappings = {
+                mappings: [
+                    null,
+                    null,
+                    [{ id: 'sra_reads', app_weight: 1, title: 'SRA Reads' }],
```
Comment (briehl) [inline test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js:63]: Added a real life mapping here to the SRA reads file to get it to show up

File: test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js [hunk 5]
```diff
@@ -155,30 +180,24 @@ define([
             ).toContain(fakeUser);
         });
 
-        it('Should show an error when a path does not exist', async () => {
+        it('Should show an error when file listing fails', async () => {
             const errorText = 'An error occurred while fetching your files';
-            jasmine.Ajax.stubRequest(/.*\/staging_service\/list\/foo?/).andReturn({
+            jasmine.Ajax.stubRequest(new RegExp(`${stagingServiceUrl}/list`)).andReturn({
```
Comment (briehl) [inline test/unit/spec/narrative_core/upload/stagingAreaViewer-spec.js:185]: No longer call `/list/` on path changes, so stub out the whole function and re-render.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 6]
```diff
@@ -102,59 +103,47 @@ define([
...
-                            [this.userInfo.user, this.subpath, f.name]
-                                .filter((p) => p.length > 0)
-                                .join('/') === f.path
-                        );
-                    });
-                    files.forEach((f) => {
-                        if (!f.isFolder) {
-                            f.imported = {};
-                        }
+                    const files = JSON.parse(data).map((f) => {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:119]: No more filtering, just get the files.
This also adds the subdirectory `subdir` to each file object.
All of the files have a `path` string which looks like:
`username/some_directory/filename.txt`

All of the staging area (and app) functions don't need the username, and that actually causes problems. So just strip it out here. This makes the path
`some_directory/filename.txt`

The `subdir`, then, becomes
`/some_directory`.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 7]
```diff
@@ -102,59 +103,47 @@ define([
...
                 })
                 .then((files) => {
                     const scrollTop = this.$elem.parent().scrollTop();
                     $(`.${fileMetadataCssBaseClass}`).detach();
-                    this.$elem.empty();
-                    this.renderFileHeader();
-                    this.renderFiles(files);
+
+                    if (this.fullDataTable) {
+                        this.fullDataTable.clear().rows.add(files).draw(false);
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:131]: From the DataTables docs: this wipes the table and adds all new data (which might be entirely different since the last refresh). `draw(false)` retains the table state - page, sorting, etc.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 8]
```diff
@@ -164,20 +153,7 @@ define([
          */
         setPath: function (path) {
             this.path = path;
-            // factor out the current subdirectory path into its own variable
-            const subpath = path.split('/');
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:168]: This was more confusing than anything, and not really helpful. It was used to figure out where files where once selected, but now the entire file path is indexed, so this doesn't matter.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 9]
```diff
@@ -271,21 +249,68 @@ define([
...
             const fileNames = stagingFiles.map((f) => f['path']);
-            const fileList = $.param({ file_list: fileNames }, true);
-            let mappings = [];
-            return Promise.resolve(
-                this.stagingServiceClient.importer_mappings({
-                    file_list: fileList,
-                })
-            )
+
+            /* Currently, the mapping info lookup is only a GET request with URL parameters.
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:263]: 😬 
I hope the staging area gets changed soon...

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 10]
```diff
@@ -309,76 +334,96 @@ define([
          */
         renderFiles: function (files) {
             const stagingAreaViewer = this;
+            const option = html.tag('option'),
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:337]: less raw HTML strings, more html function things.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 11]
```diff
@@ -309,76 +334,96 @@ define([
...
-
-            const $fileTable = $(
-                stagingAreaViewer.ftpFileTableTmpl({
-                    files: files,
-                    uploaders: stagingAreaViewer.uploaders.dropdown_order,
-                })
-            );
-
-            stagingAreaViewer.$elem.append($fileTable);
+            const emptyMsg = 'No files found.',
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:345]: Modified some text to reflect that we're looking at a file system, not just table entries.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 12]
```diff
@@ -309,76 +334,96 @@ define([
          */
         renderFiles: function (files) {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:335]: This is where most of the heaviest changes here.
Some highlights:
* file data gets fed into the table directly, not from the DOM
* file paths are used throughout, not just names
* the rowCallback function was a bit of a nightmare of nested function definitions. I decomposed that out into widget-level functions for clarity.
* now makes use of `createdCell` and `createdRow` to make DOM changes (like classes and attributes)

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 13]
```diff
@@ -439,294 +495,399 @@ define([
...
-                                stagingAreaViewer.disableImportButton();
-                            }
-                        }
+                    // attach events to the row checkbox
+                    if (stagingAreaViewer.checkedFiles.has(data.path)) {
+                        $rowCheckbox.prop('checked', true);
                     }
+                    // bind the change event here so it gets triggered
+                    // when selecting a type by the dropdown makes the change
+                    $rowCheckbox.off('click change').on('click change', (e) => {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:603]: It's a little fuzzy when and how elements get added/removed/emptied from the DOM. So we remove any events before rebinding them for the first time.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 14]
```diff
@@ -439,294 +495,399 @@ define([
...
                 } else {
-                    stagingAreaViewer.disableImportButton();
+                    $('.kb-dropzone').css('min-height', '200px');
+                    $tr.next().detach();
+                    delete this.openFileInfo[data.path];
                 }
+            });
+        },
+
+        updateTableDrawPath: function () {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:867]: This uses some DataTables trickery to auto filter to the currently set path.
Based on this plug-in info: https://datatables.net/examples/plug-ins/range_filtering.html

Basically, this adds a search filter function that iterates over all rows. Any row that returns `true` passes the filter and gets shown.

This gets updated on redraw to use whatever the path is at the time.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 15]
```diff
@@ -796,6 +951,9 @@ define([
                                 $upaField.empty();
                                 $upaField.addClass('alert alert-danger');
                                 $upaField.append(xhr.error.message);
+                            })
+                            .finally(() => {
+                                $upaField.removeClass('fa fa-spinner fa-spin');
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:956]: Goodbye spinning object name! 😢

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 16]
```diff
@@ -1015,8 +1166,13 @@ define([
                 }
 
                 if (appInfo.app_output_param) {
+                    let outputName = file;
+                    // if we're in a subpath, need to strip it down to just the file name
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:1170]: Something similar needs to be done for the bulk import setup, including adding the file type suffix. I'll try to do that next.

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeStagingDataTab.js [hunk 17]
```diff
@@ -87,8 +87,9 @@ define([
                     userInfo: userInfo,
                 });
                 this.stagingAreaViewer = new StagingAreaViewer(this.$myFiles, stagingAreaArgs);
+                this.stagingAreaViewer.render();
 
-                this.updateView();
+                // this.updateView();
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeStagingDataTab.js:92]: rm comment

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js (hunk 9 above)
Comment (ialarmedalien) [reply kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:263]: erk!!

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeStagingDataTab.js (hunk 17 above)
Comment (briehl) [reply kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeStagingDataTab.js]: done!

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js (hunk 6 above)
Comment (ialarmedalien) [reply kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:119]: much simpler!

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 18]
```diff
@@ -439,294 +495,399 @@ define([
...
+                // And if we have a suggested type, select it and enable the checkboxes
+                if (suggestedType) {
+                    importDropdown
+                        .select2({
+                            containerCssClass: `${cssBaseClass}-body__import-dropdown ${cssBaseClass}-body__import-type-selected`,
+                            placeholder: 'make the empty option disappear',
+                        })
+                        .val(suggestedType.id)
+                        .trigger('change');
+                    enableCheckboxes(suggestedType.id);
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:762]: not too major, but you could use an `else if`, which would allow you to decrease the indent level to improve readability a little.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js (hunk 14 above)
Comment (ialarmedalien) [reply kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:867]: Very cool!

It seems as though there are numerous cases where embracing DataTables fully at the start would have saved time and trouble later on. I really like DT, it's powerful and very customisable.

**Review by ialarmedalien — APPROVED** [review]
A few very minor comments -- looks good to go!

**Comment by briehl** [comment]
Thanks for the comments, I'll make that else if switch (not sure how I missed that...) and merge.
